include/onnxruntime/core/session/onnxruntime_cxx_inline.h

```diff
@@ -1996,9 +1998,10 @@ inline ShapeInferContext::ShapeInferContext(const OrtApi* ort_api,
   }
 }
 
-inline Status ShapeInferContext::SetOutputShape(size_t indice, const Shape& shape) {
+inline Status ShapeInferContext::SetOutputShape(size_t indice, const Shape& shape, ONNXTensorElementDataType type) {
   OrtTensorTypeAndShapeInfo* info = {};
   ORT_CXX_RETURN_ON_API_FAIL(ort_api_->CreateTensorTypeAndShapeInfo(&info));
+  ORT_CXX_RETURN_ON_API_FAIL(ort_api_->SetTensorElementType(info, type));
```

Comment:
I believe the underlying C API (ShapeInferContext_SetOutputTypeShape) currently ignores the tensor element type, so this change alone will not have any effect. 

We would probably need to update the C API to use the element type, and we would need to add unit tests to verify the functionality. Please refer to:
https://github.com/microsoft/onnxruntime/blob/218301403d76c3dbd7db4731071876273dfaa40d/onnxruntime/core/session/custom_ops.cc#L346-L356

https://github.com/microsoft/onnxruntime/blob/218301403d76c3dbd7db4731071876273dfaa40d/onnxruntime/core/session/custom_ops.cc#L93-L109

Reply:
Make sense ！